feat: implement open feature provider (#1)
bucketeer-io/openfeature-swift-client-sdk: README.md

# Bucketeer - OpenFeature Swift provider

This is the official Swift OpenFeature provider for accessing your feature flags with [Bucketeer](https://bucketeer.io/).

> [Bucketeer](https://bucketeer.io) is an open-source platform created by [CyberAgent](https://www.cyberagent.co.jp/en) to help teams make better decisions, reduce deployment lead time and release risk through feature flags. Bucketeer offers advanced features like dark launches and staged rollouts that perform limited releases based on user attributes, devices, and other segments.

Reviewer: ```suggestion
[Bucketeer](https://bucketeer.io) is an open-source platform created by [CyberAgent](https://www.cyberagent.co.jp/en/) to help teams make better decisions, reduce deployment lead time and release risk through feature flags. Bucketeer offers advanced features like dark launches and staged rollouts that perform limited releases based on user attributes, devices, and other segments.
```

Author: @kakcy did you set the API Key & API Domain ?

> [!WARNING]
> This is a beta version. Breaking changes may be introduced before general release.

For documentation related to flags management in Bucketeer, refer to the [Bucketeer documentation website](https://docs.bucketeer.io/sdk/client-side/ios).

## Supported iOS and Xcode versions

Minimum build tool versions:

| Tool  | Version |
| ----- | ------- |
| Xcode | 16.0+   |
| Swift | 5.0+    |

Minimum device platforms:

| Platform | Version |
| -------- | ------- |
| iOS      | 14.0    |
| tvOS     | 14.0    |

## Installation

### Swift Package Manager

With <a href="https://github.com/apple/swift-package-manager">Swift Package Manager</a>, add a <a href="https://developer.apple.com/documentation/xcode/adding_package_dependencies_to_your_app">package dependency to your Xcode project</a>. 

Select **File > Swift Packages > Add Package Dependency** and enter the repository URL: `https://github.com/bucketeer-io/openfeature-swift-client-sdk.git`.

Next select the product "BucketeerOpenFeature" and add it to your app target.

## Usuage

### Initialize the provider

> Bucketeer provider needs to be created and then set in the global OpenFeatureAPI. 

Reviewer 2: ```suggestion
Bucketeer provider needs to be created and then set in the global OpenFeatureAPI.
```

```swift
import BucketeerOpenFeature
import OpenFeature
import Bucketeer

do {
  // SDK configuration
  let userId = "targetingUserId"
  let config = try BKTConfig.Builder()
    .with(apiKey: "YOUR_API_KEY")
    .with(apiEndpoint: "YOUR_API_ENDPOINT")
    .with(featureTag: "YOUR_FEATURE_TAG")
    .with(appVersion: Bundle.main.infoDictionary?["CFBundleShortVersionString"] as! String)
    .build()
    let provider = BucketeerProvider(config: config)
    // User configuration
    let context = MutableContext(
        targetingKey: userId,
        structure: MutableStructure(
            attributes: [:] // The user attributes are optional
        )
    )

    // Set context before initialize
    await OpenFeatureAPI.shared.setProviderAndWait(provider: provider, initialContext: context)
} catch {
  // Error handling
}
```

> See our [documentation](https://docs.bucketeer.io/sdk/client-side/ios) for more SDK configuration..
> 
> The evaluation context is the way for the client to specify contextual data that Bucketeer uses to evaluate the feature flags.
> 
> The `targetingKey` should be the id of a user and should not be empty string.
> 
> ### Update the Evaluation Context
> 
> If the user attributes change, you can update the evaluation context with the new attributes.

Reviewer 2: ```suggestion
See our [documentation](https://docs.bucketeer.io/sdk/client-side/ios) for more SDK configuration..

The evaluation context allows the client to specify contextual data that Bucketeer uses to evaluate the feature flags.

The `targetingKey` is the user ID (Unique ID) and cannot be empty.

### Update the Evaluation Context

You can update the evaluation context with the new attributes if the user attributes change.
```

```swift
let ctx = MutableContext(targetingKey: userId,  structure: MutableStructure(
            attributes: ["buyer":"true"]
        ))
OpenFeatureAPI.shared.setEvaluationContext(evaluationContext: ctx)
```

> [!WARNING]
> Changing the `targetingKey` is not supported in the current implementation of the BucketeerProvider.

> To handle a user ID change, the BucketeerProvider must be remove and reinitialized.

Reviewer 2: ```suggestion
To change the user ID, the BucketeerProvider must be removed and reinitialized.
```
